File: library/src/main/java/com/thesurix/gesturerecycler/GestureTouchHelperCallback.java
package com.thesurix.gesturerecycler;

import android.graphics.Canvas;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;

import static com.thesurix.gesturerecycler.LayoutFlags.GRID;
import static com.thesurix.gesturerecycler.LayoutFlags.LINEAR;
import static com.thesurix.gesturerecycler.LayoutFlags.STAGGERED;

/**
 * Touch helper callback that handles different RecycleView gestures.
 * @author thesurix
 */
public class GestureTouchHelperCallback extends ItemTouchHelper.Callback {

    /** GestureAdapter instance for touch callbacks */
    private final GestureAdapter mGestureAdapter;

    /** Flag that enables or disables swipe gesture */
    private boolean mIsSwipeEnabled;
    /** Flag that enables or disables manual drag gesture */
    private boolean mIsManualDragEnabled;
    /** Flag that enables long press drag gesture */
    private boolean mIsLongPressDragEnabled;

    /** Flags for drag gesture */
    private int mDragFlags = ItemTouchHelper.UP | ItemTouchHelper.DOWN;
    /** Flags for swipe gesture */
    private int mSwipeFlags = ItemTouchHelper.RIGHT;

    /**
     * Constructs callback object based on passed adapter.
     * @param adapter adapter
     */
    public GestureTouchHelperCallback(final GestureAdapter adapter) {
        mGestureAdapter = adapter;
    }

    @Override
    public int getMovementFlags(final RecyclerView recyclerView, final RecyclerView.ViewHolder viewHolder) {
        final GestureViewHolder holder = (GestureViewHolder) viewHolder;
        return makeMovementFlags(holder.canDrag() ? mDragFlags : 0, holder.canSwipe() ? mSwipeFlags : 0);
    }

    @Override
    public boolean onMove(final RecyclerView recyclerView, final RecyclerView.ViewHolder source, final RecyclerView.ViewHolder target) {
        return mGestureAdapter.onItemMove(source.getAdapterPosition(), target.getAdapterPosition());
    }

    @Override
    public void onSwiped(final RecyclerView.ViewHolder viewHolder, final int direction) {
        mGestureAdapter.onItemDismissed(viewHolder.getAdapterPosition());
    }

    @Override
    public void onSelectedChanged(final RecyclerView.ViewHolder viewHolder, final int actionState) {
        super.onSelectedChanged(viewHolder, actionState);
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE && viewHolder instanceof GestureViewHolder) {
            final GestureViewHolder itemViewHolder = (GestureViewHolder) viewHolder;
            final View backgroundView = itemViewHolder.getBackgroundView();
            if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE && backgroundView != null) {
                backgroundView.setVisibility(View.VISIBLE);
            }

            itemViewHolder.onItemSelect();
        }
    }

    @Override
    public void onChildDraw(final Canvas c, final RecyclerView recyclerView, final RecyclerView.ViewHolder viewHolder, final float dX, final float dY,
            final int actionState, final boolean isCurrentlyActive) {
        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
            final View foregroundView = ((GestureViewHolder) viewHolder).getForegroundView();
            getDefaultUIUtil().onDraw(c, recyclerView, foregroundView, dX, dY, actionState, isCurrentlyActive);
        } else {
            super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
        }
    }

    @Override
    public void clearView(final RecyclerView recyclerView, final RecyclerView.ViewHolder viewHolder) {
        super.clearView(recyclerView, viewHolder);
        mGestureAdapter.onItemMoved();
        if (viewHolder instanceof GestureViewHolder) {
            final GestureViewHolder itemViewHolder = (GestureViewHolder) viewHolder;
            itemViewHolder.onItemClear();

            final View backgroundView = itemViewHolder.getBackgroundView();
            if (backgroundView != null) {
                backgroundView.setVisibility(View.GONE);
            }

            final View foregroundView = itemViewHolder.getForegroundView();
            getDefaultUIUtil().clearView(foregroundView);
        }
    }

    @Override
    public boolean isLongPressDragEnabled() {
        return mIsLongPressDragEnabled;
    }

    @Override
    public boolean isItemViewSwipeEnabled() {
        return mIsSwipeEnabled;
    }

    /**
     * Returns true if manual drag is enabled, false if manual drag is disabled.
     * @return manual drag state
     */
    public boolean isManualDragEnabled() {
        return mIsManualDragEnabled;
    }

    /**
     * Sets manual drag gesture enabled or disabled.
     * Manual drag is disabled by default.
     * @param enabled true to enable, false to disable
     */
    public void setManualDragEnabled(final boolean enabled) {
        mIsManualDragEnabled = enabled;
        //update adapter to enable/disable manual drag spawns
        mGestureAdapter.allowManualDrag(mIsManualDragEnabled);
    }

    /**
     * Sets swipe gesture enabled or disabled.
     * Swipe is disabled by default.
     * @param enabled true to enable, false to disable
     */
    public void setSwipeEnabled(final boolean enabled) {
        mIsSwipeEnabled = enabled;
    }

    /**
     * Sets long press drag gesture enabled or disabled.
     * Long press drag is disabled by default.
     * @param enabled true to enable, false to disable
     */
    public void setLongPressDragEnabled(final boolean enabled) {
        mIsLongPressDragEnabled = enabled;
    }

    /**
     * Sets predefined drag flags for RecyclerView layout.
     * @param layout type of the RecyclerView layout
     */
    public void setDragFlagsForLayout(final RecyclerView.LayoutManager layout) {
        if (layout instanceof GridLayoutManager) {
            mDragFlags = GRID.getDragFlags(layout);
        } else if (layout instanceof LinearLayoutManager) {
            mDragFlags = LINEAR.getDragFlags(layout);
        } else if (layout instanceof StaggeredGridLayoutManager) {
            mDragFlags = STAGGERED.getDragFlags(layout);
        } else {
            throw new IllegalArgumentException("Unsupported layout type.");
        }
    }

    /**
     * Sets predefined swipe flags for RecyclerView layout.
     * @param layout type of the RecyclerView layout
     */
    public void setSwipeFlagsForLayout(final RecyclerView.LayoutManager layout) {
        if (layout instanceof GridLayoutManager) {
            mSwipeFlags = GRID.getSwipeFlags(layout);
        } else if (layout instanceof LinearLayoutManager) {
            mSwipeFlags = LINEAR.getSwipeFlags(layout);
        } else if (layout instanceof StaggeredGridLayoutManager) {
            mSwipeFlags = STAGGERED.getSwipeFlags(layout);
        } else {
            throw new IllegalArgumentException("Unsupported layout type.");
        }
    }

    /**
     * Sets flags for drag gesture.
     * See {@link ItemTouchHelper} flags.
     * @param flags flags for drag gesture
     */
    public void setDragFlags(final int flags) {
        mDragFlags = flags;
    }

    /**
     * Sets flags for swipe gesture.
     * See {@link ItemTouchHelper} flags.
     * @param flags flags for swipe gesture
     */
    public void setSwipeFlags(final int flags) {
        mSwipeFlags = flags;
    }
}

File: library/src/main/java/com/thesurix/gesturerecycler/GestureTouchHelperCallback.kt
package com.thesurix.gesturerecycler

import android.graphics.Canvas
import android.util.Log
import android.view.View
import androidx.recyclerview.widget.*
import com.thesurix.gesturerecycler.LayoutFlags.*

/**
 * Touch helper callback that handles different RecycleView gestures.
 * Constructs callback object based on passed adapter.
 * @param adapter adapter
 * @author thesurix
 */
private val DIRECTIONS = listOf(ItemTouchHelper.LEFT, ItemTouchHelper.RIGHT, ItemTouchHelper.UP, ItemTouchHelper.DOWN)
class GestureTouchHelperCallback(private val gestureAdapter: GestureAdapter<*, *>) : ItemTouchHelper.Callback() {

    /** Flag that enables or disables swipe gesture  */
    var swipeEnabled = false
    /** Flag that enables or disables manual drag gesture  */
    var manualDragEnabled = false
        set(enabled) {
            field = enabled
            gestureAdapter.allowManualDrag(manualDragEnabled)
        }
    /** Flag that enables long press drag gesture  */
    var longPressDragEnabled = false

    /** Flags for drag gesture  */
    var dragFlags = ItemTouchHelper.UP or ItemTouchHelper.DOWN
    /** Flags for swipe gesture  */
    var swipeFlags = ItemTouchHelper.RIGHT

    override fun getMovementFlags(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder): Int {
        val holder = viewHolder as GestureViewHolder<*>
        return makeMovementFlags(if (holder.canDrag()) dragFlags else 0, if (holder.canSwipe()) swipeFlags else 0)
    }

    override fun onMove(recyclerView: RecyclerView, source: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
        return gestureAdapter.onItemMove(source.adapterPosition, target.adapterPosition)
    }

    override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
        gestureAdapter.onItemDismissed(viewHolder.adapterPosition, direction)
    }

    override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
        super.onSelectedChanged(viewHolder, actionState)
        if (actionState != ItemTouchHelper.ACTION_STATE_IDLE && viewHolder is GestureViewHolder<*>) {
            viewHolder.onItemSelect()
        }
    }

    override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float,
                             actionState: Int, isCurrentlyActive: Boolean) {
        when(actionState) {
            ItemTouchHelper.ACTION_STATE_SWIPE -> {
                val direction = when {
                    dX.compareTo(0f) == 0 -> if (dY < 0) ItemTouchHelper.UP else ItemTouchHelper.DOWN
                    dY.compareTo(0f) == 0 -> if (dX < 0) ItemTouchHelper.LEFT else ItemTouchHelper.RIGHT
                    else -> -1
                }

                val gestureViewHolder = (viewHolder as GestureViewHolder<*>)
                hideBackgroundViews(gestureViewHolder)

                if (direction != -1) {
                    val backgroundView = gestureViewHolder.getBackgroundView(direction)
                    backgroundView?.let {
                        if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE && backgroundView.visibility == View.GONE) {
                            backgroundView.visibility = View.VISIBLE
                        }
                    }
                }

                val foregroundView = gestureViewHolder.foregroundView
                getDefaultUIUtil().onDraw(c, recyclerView, foregroundView, dX, dY, actionState, isCurrentlyActive)
            }
            else -> super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
        }
    }

    override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
        super.clearView(recyclerView, viewHolder)
        gestureAdapter.onItemMoved()
        if (viewHolder is GestureViewHolder<*>) {
            viewHolder.onItemClear()
            hideBackgroundViews(viewHolder)

            val foregroundView = viewHolder.foregroundView
            getDefaultUIUtil().clearView(foregroundView)
        }
    }

    private fun hideBackgroundViews(viewHolder: GestureViewHolder<*>) {
        DIRECTIONS.forEach {
            viewHolder.getBackgroundView(it)?.visibility = View.GONE
        }
    }

    override fun isLongPressDragEnabled(): Boolean {
        return longPressDragEnabled
    }

    override fun isItemViewSwipeEnabled(): Boolean {
        return swipeEnabled
    }

    /**
     * Sets predefined drag flags for RecyclerView layout.
     * @param layout type of the RecyclerView layout
     */
    fun setDragFlagsForLayout(layout: RecyclerView.LayoutManager) {
        dragFlags = when (layout) {
            is GridLayoutManager -> GRID.getDragFlags(layout)
            is LinearLayoutManager -> LINEAR.getDragFlags(layout)
            is StaggeredGridLayoutManager -> STAGGERED.getDragFlags(layout)
            else -> throw IllegalArgumentException("Unsupported layout type.")
        }
    }

    /**
     * Sets predefined swipe flags for RecyclerView layout.
     * @param layout type of the RecyclerView layout
     */
    fun setSwipeFlagsForLayout(layout: RecyclerView.LayoutManager) {
        swipeFlags = when (layout) {
            is GridLayoutManager -> GRID.getSwipeFlags(layout)
            is LinearLayoutManager -> LINEAR.getSwipeFlags(layout)
            is StaggeredGridLayoutManager -> STAGGERED.getSwipeFlags(layout)
            else -> throw IllegalArgumentException("Unsupported layout type.")
        }
    }
}
